### PR TITLE
Postgres SELECT ... FOR UPDATE Support

### PR DESCRIPTION
**Pull request by SamProtas**
Referencing: https://github.com/tathougies/beam/issues/152

### CODE REVIEW

**Comment by SamProtas** [comment]
I _think_ I've got the various locking clause forms covered. Although I was unsure of how to specify the table names for the `OF <table1>[,<table2>,...]` exactly.

**Comment by SamProtas** [comment]
It's likely I'm missing some things, but at this point I'm thinking the following:

- We can't use `buildSql92Query` anymore for postgres' `HasQBuilder` instance
    - I'm guessing this is "correct" in the sense that this functionality isn't in the 92 standard
    - Maybe some (most) of this convenience function can be factored out to be reusable for a `buildPostgreSqlQuery` function 
    - If my next thought below is true, maybe a `buildQueryExtensions` function can be passed in to share most of this logic still
- `QF` may need to be extended in some way. I'm not sure the best way to do this
    - Maybe we could add another type parameter (defaulted to `()`) and constructor for backends to included a nested free-monad defining their additons (I'm not 100% sure this is even possible).

Or maybe there's something much simpler I haven't figured out.

File: beam-postgres/Database/Beam/Postgres/Syntax.hs [hunk 1]
```diff
@@ -1371,6 +1421,23 @@ pgDropExtensionSyntax extName =
 --                                        Nothing
 --      buildJoinFrom tbl newSource Nothing
 
+
+-- TODO
+-- Don't directly expose sum-type constructors?
```
Comment (tathougies) [inline beam-postgres/Database/Beam/Postgres/Syntax.hs]: They should be exposed from the `Syntax` module, but they should not be exposed from the main `Database.Beam.Postgres` module, if that makes sense.

File: beam-postgres/Database/Beam/Postgres/Syntax.hs [hunk 2]
```diff
@@ -1371,6 +1421,23 @@ pgDropExtensionSyntax extName =
...
+-- Constrain 'table' to exist within the query (using s maybe?)
+-- Constrain Q to not have aggregations (at least the locked tables I think, using s?)
+--
+--   https://www.postgresql.org/docs/10/static/sql-select.html#SQL-FOR-UPDATE-SHARE
+--   "The locking clauses cannot be used in contexts where returned rows cannot be clearly
+--   identified with individual table rows; for example they cannot be used with aggregation."
```
Comment (tathougies) [inline beam-postgres/Database/Beam/Postgres/Syntax.hs]: Oh this is an interesting restriction. I'll have to think what the best way to deal with this one is. We could probably do it with the `s` parameter, but may need to make some changes in `beam-core`.

File: beam-postgres/Database/Beam/Postgres/Syntax.hs [hunk 3]
```diff
@@ -1371,6 +1421,23 @@ pgDropExtensionSyntax extName =
 --                                        Nothing
 --      buildJoinFrom tbl newSource Nothing
 
+
+-- TODO
+-- Don't directly expose sum-type constructors?
+-- Constrain 'table' to exist within the query (using s maybe?)
```
Comment (tathougies) [inline beam-postgres/Database/Beam/Postgres/Syntax.hs]: I think we can probably do some trickery with existential types on `db` to make sure this works. Let me think about it.

**Comment by tathougies** [comment]
@SamProtas Overall, I'm happy accepting this pull request as is and then work on these issues one by one. If you could just modify it so the constructors are exposed, that'd be great.

**Comment by SamProtas** [comment]
Let me know if you had something else in mind.

**Comment by tathougies** [comment]
Perfect! Thank you very much. Please add yourself to the CONTRIBUTORS file, if you've not already done so!

**Pull request merged**